### PR TITLE
`Learning path`: Refactor learning path navigation overview

### PR DESCRIPTION
**Pull request by JohannesStoehr**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

The navigation overview (where all the learning objects of the learning path are shown) should be grouped by competency.

### Description

The navigation overview now shows a list of the competencies of the learning path. Each competency can be expanded to show its corresponding learning objects. These objects can be selected to navigate to that specific learning objects.
To achieve these client tests two additional endpoints have been added on the server.
Stacked on #8733 

Testing this PR is straightforward on TS5 (Course Johannes Wiest) because the necessary competencies and other elements are already set up.

- one to fetch all competencies of a learning path
- one to fetch all learning objects of a competency

### Steps for Testing

Prerequisites:
- 1 Student
- 1 Course
- at least 1 competency with linked learning objects (LectureUnit or Exercise)

1. Log in to Artemis
2. Navigate to Course
3. Navigate to Learning path
4. Navigate back and forth in learning path (Previous and Next-Button)
5. Open Competency overview dropdown
6. Open Competency learning objects accordion
7. Navigate to a learning object by clicking on an item

### Info for code review
Due to merging develop into this branch the changes look more than there are from this PR. This PR includes more around (+900/-300) lines and changed the following files:

**Server changes:**
repository/CompetencyProgressRepository.java
repository/CompetencyRepository.java
repository/CourseCompetencyRepository.java
service/LearningObjectService.java
service/competency/CompetencyProgressService.java
service/learningpath/LearningPathNavigationService.java
service/learningpath/LearningPathRecommendationService.java
web/rest/LearningPathResource.java
web/rest/dto/competency/LearningPathNavigationObjectDTO.java

**Server tests:**
src/test/java/de/tum/in/www1/artemis/competency/LearningPathIntegrationTest.java

**Client changes:**
course/learning-paths/components/competency-graph-modal/competency-graph-modal.component.ts
course/learning-paths/components/competency-graph/competency-graph.component.ts
course/learning-paths/components/competency-node/competency-node.component.ts
course/learning-paths/components/learning-path-lecture-unit/learning-path-lecture-unit.component.ts
course/learning-paths/components/learning-path-nav-overview-learning-objects/learning-path-nav-overview-learning-objects.component.html
course/learning-paths/components/learning-path-nav-overview-learning-objects/learning-path-nav-overview-learning-objects.component.ts
course/learning-paths/components/learning-path-nav-overview-learning-objects/learning-path-nav-overview-learning-objects.component.scss
course/learning-paths/components/learning-path-nav-overview/learning-path-nav-overview.component.html
course/learning-paths/components/learning-path-nav-overview/learning-path-nav-overview.component.scss
course/learning-paths/components/learning-path-nav-overview/learning-path-nav-overview.component.ts
course/learning-paths/components/learning-path-student-nav/learning-path-student-nav.component.html
course/learning-paths/components/learning-path-student-nav/learning-path-student-nav.component.ts
course/learning-paths/pages/learning-path-student-page/learning-path-student-page.component.ts
course/learning-paths/services/base-api-http.service.ts
course/learning-paths/services/learning-path-api.service.ts
course/learning-paths/services/learning-path-navigation.service.ts
entities/competency/learning-path.model.ts

**Client tests:**
spec/component/learning-paths/components/learning-path-nav-overview-learning-objects.component.spec.ts
spec/component/learning-paths/components/learning-path-nav-overview.component.spec.ts
spec/component/learning-paths/components/learning-path-nav.component.spec.ts
spec/component/learning-paths/pages/learning-path-student-page.component.spec.ts
spec/service/learning-path/learning-path-api.service.spec.ts

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| competency-form.component.ts | 87.77% | ✅ |
| competency-graph-modal.component.ts | 90.9% | ✅ |
| competency-graph.component.ts | 85.29% | ✅ |
| competency-node.component.ts | 100% | ✅ |
| learning-path-lecture-unit.component.ts | 81.48% | ✅ |
| learning-path-nav-overview-learning-objects.component.ts | 90% | ✅ |
| learning-path-nav-overview.component.ts | 90.62% | ✅ |
| learning-path-student-nav.component.ts | 100% | ✅ |
| learning-path.service.ts | 75% | ✅ |
| learning-path-student-page.component.ts | 97.43% | ✅ |
| learning-path-container.component.ts | 87% | ✅ |
| base-api-http.service.ts | 72.22% | ✅ |
| learning-path-api.service.ts | 100% | ✅ |
| learning-path-navigation.service.ts | 100% | ✅ |
| learning-path.model.ts | 96.15% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| Exercise.java | 91% | ✅ |
| LectureUnit.java | 100% | ✅ |
| LearningObjectService.java | 89% | ✅ |
| CompetencyProgressService.java | 94% | ✅ |
| LearningPathNavigationService.java | 87% | ✅ |
| LearningPathNgxService.java | 97% | ✅ |
| LearningPathRecommendationService.java | 97% | ✅ |
| LearningPathService.java | 97% | ✅ |
| LearningPathResource.java | 100% | ✅ |

### Screenshots

image

### CODE REVIEW

**Review by Jan-Thurner — CHANGES_REQUESTED** [review]
When trying to access the learning paths, an internal server error is thrown and the page is empty: 
image

Edit: The newest push has not yet been deployed. I am retesting now.

**Review by Jan-Thurner — APPROVED** [review]
Tested again on TS5. Works as expected.

File: src/test/javascript/spec/component/learning-paths/components/learning-path-nav.component.spec.ts [hunk 1]
```diff
@@ -175,10 +175,10 @@ describe('LearningPathStudentNavComponent', () => {
         const navOverviewButton = fixture.debugElement.query(By.css('#navigation-overview'));
         navOverviewButton.nativeElement.click();
         fixture.detectChanges();
-        const navOverview = fixture.debugElement.query(By.directive(LearningPathStudentNavOverviewComponent));
+        const navOverview = fixture.debugElement.query(By.directive(LearningPathNavOverviewComponent));
         expect(navOverview).toBeTruthy();
         expect(setShowNavigationOverviewSpy).toHaveBeenCalledWith(true);
-        expect(component.showNavigationOverview()).toBeTrue();
+        expect(loadCompetenciesSpy).toHaveBeenCalledWith(learningPathId);
```
Comment (pzdr7) [inline src/test/javascript/spec/component/learning-paths/components/learning-path-nav.component.spec.ts:181]: Exactly once?

**Review by pzdr7 — APPROVED** [review]

**Review by rstief — APPROVED** [review]
Tested on ts5 LGTM. 
Just a minor nitpick: The `accordion-body` styles set the `overflow` to `scroll` which causes it to look like this on windows:
image

**Review by EneaGore — COMMENTED** [review]
Testing on TS5. Works as expected, but after i delete an exercise which was connected to a competency i get an error 500 when accessing the page again. Probably its not specifically connected to this PR though.
500nav

**Review by edkaya — APPROVED** [review]
Tested on ts5, works as expected. Some UI sidenotes that can be done in a follow up PR: 
- I see a scrollbar in content part, which can be hidden if you don't have an exceeding content. 
Screenshot 2024-06-25 at 19 49 32
- We might want to have the overview scrollbar in auto mode instead scroll 
Screenshot 2024-06-25 at 19 50 38

**Comment by JohannesWt** [comment]
> Tested on ts5, works as expected. Some UI sidenotes that can be done in a follow up PR:
> 
> * I see a scrollbar in content part, which can be hidden if you don't have an exceeding content.
> 
> Screenshot 2024-06-25 at 19 49 32 - We might want to have the overview scrollbar in auto mode instead scroll Screenshot 2024-06-25 at 19 50 38

Unfortunately we have to use scroll overview here. If we use "auto" the dropdown will grow indefinitely in size ignoring the max-height. The horizontal scroll in the content part is already documented in an issue and will be solved in a follow up PR.

**Comment by az108** [comment]
I like the changes, maybe in a future PR we could show somewhere which compentency one is at rn. Because right now I would have to click on the dropdown to see that i am in competency 2. 
image

**Comment by JohannesWt** [comment]
> I like the changes, maybe in a future PR we could show somewhere which compentency one is at rn. Because right now I would have to click on the dropdown to see that i am in competency 2. 😄 image

That's exactly the plan. In the future the current competency should be marked and opened, as soon as you open the drop down.

**Review by MaximilianAnzinger — APPROVED** [review]

**Pull request merged**